File: build-single-file.pl
#!/usr/bin/perl
# Fallback for build-single-file.py — does exactly the same job, but with
# Perl instead of Python. Use this if python3 refuses to run with:
#
#     You have not agreed to the Xcode license agreements.
#
# That's a one-time macOS prompt, not a bug in either script — fix it
# permanently with:
#
#     sudo xcodebuild -license
#
# and go back to `python3 build-single-file.py` afterwards. Until then, or if
# you'd rather not run a sudo command, this does the same inlining with the
# Perl that ships with every Mac — nothing to install.
#
#     perl build-single-file.pl

use strict;
use warnings;
use MIME::Base64 qw(encode_base64);
use FindBin qw($RealBin);

my $root   = $RealBin;
my $source = "$root/index.html";
my $outdir = "$root/dist";
my $out    = "$outdir/index.html";

die "index.html not found at $source\n" unless -f $source;

open(my $ifh, "<:raw", $source) or die "$!";
local $/;
my $html = <$ifh>;
close $ifh;
printf STDERR "Reading index.html (%d KB)\n", length($html) / 1024;

my %mime_for = (
    png => "image/png", jpg => "image/jpeg", jpeg => "image/jpeg",
    svg => "image/svg+xml",
);
my $inlined = 0;

# Minification happens only here, in the bundle — index.html itself keeps
# every comment and its one-selector-per-line formatting, since that's the
# file people actually edit. This just strips the same bytes back out of the
# copy that gets deployed.
#
# CSS gets a real minify: strip /* */ comments, collapse whitespace, trim
# around { } ;. Safe because the stylesheet has no url(...) values — a colon
# or slash inside one of those is exactly the kind of thing a careless
# minifier corrupts, so that's checked before this ships, not assumed.
sub minify_css {
    my ($css) = @_;
    $css =~ s{/\*.*?\*/}{}gs;
    $css =~ s{[ \t\r\n]+}{ }g;
    $css =~ s{ *([{};]) *}{$1}g;
    $css =~ s{^\s+|\s+$}{}g;
    return $css;
}

# JS gets a deliberately smaller pass: only /* */ block comments and blank
# lines. Line comments (//) are left alone — a regex can't reliably tell a
# real // comment from one that's part of a string or a `https://` URL
# without actually tokenising the code, and getting that wrong silently
# breaks the page. The full block-comment style used throughout this file's
# script is still worth stripping; it's most of the saving anyway.
sub minify_js {
    my ($js) = @_;
    $js =~ s{/\*.*?\*/}{}gs;
    $js =~ s{\n[ \t]*\n}{\n}g;
    $js =~ s{^[ \t]+}{}gm;
    $js =~ s{^\s+|\s+$}{}g;
    return $js;
}

my ($css_before, $css_after) = (0, 0);
$html =~ s{(<style>)(.*?)(</style>)}{
    $css_before += length($2);
    my $min = minify_css($2);
    $css_after += length($min);
    "$1$min$3";
}gse;

my ($js_before, $js_after) = (0, 0);
$html =~ s{(<script>)(.*?)(</script>)}{
    $js_before += length($2);
    my $min = minify_js($2);
    $js_after += length($min);
    "$1$min$3";
}gse;

if ($css_before) {
    printf STDERR "Minified CSS: %d KB -> %d KB\n", $css_before / 1024, $css_after / 1024;
}
if ($js_before) {
    printf STDERR "Minified JS:  %d KB -> %d KB\n", $js_before / 1024, $js_after / 1024;
}

# Matches both src="assets/…" (images) and href="assets/…" (favicon <link>
# tags) so the single-file bundle is genuinely self-contained either way.
$html =~ s{(src=|href=)(["'])(assets/[^"']+)\2}{
    my ($attr, $quote, $rel) = ($1, $2, $3);
    my $path = "$root/$rel";

    if (-f $path) {
        open(my $afh, "<:raw", $path) or die "cannot read $path: $!";
        local $/;
        my $bytes = <$afh>;
        close $afh;

        my ($ext) = $rel =~ /\.([^.]+)$/;
        my $mime = $mime_for{lc($ext // '')} // "application/octet-stream";
        my $b64 = encode_base64($bytes, "");

        $inlined++;
        printf STDERR "  inlined: %s (%d KB)\n", $rel, length($bytes) / 1024;
        qq{${attr}${quote}data:$mime;base64,$b64${quote}};
    } else {
        print STDERR "  skip (missing): $rel\n";
        $&;
    }
}ge;

mkdir $outdir unless -d $outdir;
open(my $ofh, ">:raw", $out) or die "$!";
print $ofh $html;
close $ofh;

printf STDERR "\nWrote dist/index.html (%d KB), %d image(s) inlined\n",
    length($html) / 1024, $inlined;

File: build-single-file.py
#!/usr/bin/env python3
"""
Inline every assets/ image referenced by index.html as a data URI and write
dist/index.html — one self-contained file that works with no assets folder.

    python3 build-single-file.py

The normal deploy target is index.html + assets/ (smaller, cacheable). Use this
when you need a single file to email, preview, or drop somewhere without a
folder structure.
"""

import base64
import mimetypes
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parent
SOURCE = ROOT / "index.html"
OUT_DIR = ROOT / "dist"
OUT = OUT_DIR / "index.html"

# Matches src="assets/whatever.png" (images) and href="assets/whatever.svg"
# (favicon <link> tags), in either quote style
ASSET_REF = re.compile(r'(src=|href=)(["\'])(assets/[^"\']+)\2')


def inline(match: re.Match) -> str:
    attr, quote, rel = match.group(1), match.group(2), match.group(3)
    path = ROOT / rel

    if not path.is_file():
        print(f"  skip (missing): {rel}")
        return match.group(0)

    mime = mimetypes.guess_type(path.name)[0] or "application/octet-stream"
    encoded = base64.b64encode(path.read_bytes()).decode("ascii")
    print(f"  inlined: {rel}  ({path.stat().st_size / 1024:.0f} KB)")
    return f"{attr}{quote}data:{mime};base64,{encoded}{quote}"


STYLE_BLOCK = re.compile(r"(<style>)(.*?)(</style>)", re.S)
SCRIPT_BLOCK = re.compile(r"(<script>)(.*?)(</script>)", re.S)


def minify_css(css: str) -> str:
    """Strip comments and collapse whitespace. Only used on the bundle — the
    source index.html keeps every comment, since that's the file people
    actually edit. Safe to be this aggressive because the stylesheet has no
    url(...) values; a colon or slash inside one of those is exactly what a
    careless minifier corrupts, so that gets checked before this ships, not
    assumed."""
    css = re.sub(r"/\*.*?\*/", "", css, flags=re.S)
    css = re.sub(r"[ \t\r\n]+", " ", css)
    css = re.sub(r" *([{};]) *", r"\1", css)
    return css.strip()


def minify_js(js: str) -> str:
    """Deliberately smaller than the CSS pass: only /* */ block comments and
    blank lines. Line comments (//) are left alone — a regex can't reliably
    tell a real // comment from one inside a string or a `https://` URL
    without actually tokenising the code, and getting that wrong silently
    breaks the page. The block-comment style used throughout this file's
    script is still most of the saving anyway."""
    js = re.sub(r"/\*.*?\*/", "", js, flags=re.S)
    js = re.sub(r"\n[ \t]*\n", "\n", js)
    js = re.sub(r"^[ \t]+", "", js, flags=re.M)
    return js.strip()


def main() -> None:
    if not SOURCE.is_file():
        raise SystemExit(f"index.html not found at {SOURCE}")

    html = SOURCE.read_text(encoding="utf-8")
    print(f"Reading {SOURCE.name} ({len(html) / 1024:.0f} KB)")

    built = ASSET_REF.sub(inline, html)

    css_before = css_after = js_before = js_after = 0

    def do_css(m: re.Match) -> str:
        nonlocal css_before, css_after
        css_before += len(m.group(2))
        minified = minify_css(m.group(2))
        css_after += len(minified)
        return f"{m.group(1)}{minified}{m.group(3)}"

    def do_js(m: re.Match) -> str:
        nonlocal js_before, js_after
        js_before += len(m.group(2))
        minified = minify_js(m.group(2))
        js_after += len(minified)
        return f"{m.group(1)}{minified}{m.group(3)}"

    built = STYLE_BLOCK.sub(do_css, built)
    built = SCRIPT_BLOCK.sub(do_js, built)

    if css_before:
        print(f"Minified CSS: {css_before / 1024:.0f} KB -> {css_after / 1024:.0f} KB")
    if js_before:
        print(f"Minified JS:  {js_before / 1024:.0f} KB -> {js_after / 1024:.0f} KB")

    OUT_DIR.mkdir(exist_ok=True)
    OUT.write_text(built, encoding="utf-8")
    print(f"\nWrote {OUT.relative_to(ROOT)} ({len(built) / 1024:.0f} KB)")


if __name__ == "__main__":
    main()
